staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go

```diff
@@ -83,31 +117,64 @@ func NewClient(urls []string, username, password string, headers map[string]stri
 }
 
 const (
-	servicePath  = "/nacos/v1/ns/service/list?pageNo=1&pageSize=100000"
-	intancesPath = "/nacos/v1/ns/instance/list?serviceName="
-	loginPath    = "/nacos/v1/auth/login"
+	namespaceListAPI      = "/nacos/v1/console/namespaces"
+	serviceListAPI        = "/nacos/v1/ns/service/list"
+	catalogServiceListAPI = "/nacos/v1/ns/catalog/services"
+	intancesListAPI       = "/nacos/v1/ns/instance/list"
+	loginAPI              = "/nacos/v1/auth/login"
 )
 
-func (c *client) chooseURL() string {
-	if c.index >= len(c.urls) {
-		c.index = 0
+func encodeQuery(param map[string]string) string {
+	if len(param) == 0 {
+		return ""
+	}
+	enc := url.Values{}
+	for k, v := range param {
+		enc.Add(k, v)
 	}
-	url := c.urls[c.index]
-	c.index++
+	return enc.Encode()
+}
 
-	return url
+func (c *client) call(api string, method string, header map[string]string, queryParam map[string]string, body io.Reader) ([]byte, error) {
+	query := encodeQuery(queryParam)
+	appendUrl := func(url string) string {
+		if query == "" {
+			return url
+		}
+		return url + "?" + query
+	}
+	var lastErr error
+	l := len(c.urls)
+	shift := rand.Intn(l)
+	for i := 0; i < l; i++ {
+		var curBody io.Reader
+		var copiedBody bytes.Buffer
```

Comment:
a little too complicated. 

as at least we should cache the whole body for once, we can read it into a buf and then start the call-loop.

Reply:
copied and create new Reader for every call